### PR TITLE
Add GroupBy option to snapshots command

### LINKED ISSUE #2037: --group-by option for snapshots command
Output of `restic version`
--------------------------

restic compiled manually compiled with go1.10.3 on freebsd/amd64

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

What are you trying to do?
--------------------------

it would be great to have the --group-by option available for the snapshots command as well as the forget command. It makes it much easier to see the sets of backups marked with tags.

`restic snapshots -g tags`

Similar grouped output as you see when you run `restic forget -g tags ...`

```
snapshots for (tags [monthly]):
      ...
snapshots for (tags [daily]):
      ...
```

Ref: https://forum.restic.net/t/feature-group-by-option-for-snapshots-command/1070/4

Did restic help you or made you happy in any way?
-------------------------------------------------

It makes me happy every day to know my backups are safe and secured :)

### PR DESCRIPTION
**Pull request by ArcticXWolf**
What is the purpose of this change? What does it change?
--------------------------------------------------------
This commit adds a --group-by option to the snapshots command, which
behaves similar to the --group-by option of forget. Valid option values
are "host, paths, tags". If this option is given, the output of
snapshots will be divided into multiple tables, according to the value
given (i.e. "host" will create a table of snapshots for each host, that
has a snapshot in the list). Also the JSON output will be grouped.

The default behavior (when --group-by is not given) has not changed.

More to this discussion can be found in issue #2037.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------
Yes, it was discussed in issue #2037.

Closes #2037

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have added tests for all changes in this PR
- [x] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Comment by ArcticXWolf** [comment]
Hey, I want to work on the group-by option discussed in #2037. The main functionality is already done, whats missing are tests, documentation and changelogs.

As this is my first contribution to restic and also my first time coding in Go, I have some questions:

1. What's used to do the testing? (As I'm coming from Ruby, I know that there are many testing frameworks/DSLs)
2. Is the style of my changes okay? I'm not sure, if I violated any naming conventions or else.

I hope my contribution helps!

File: cmd/restic/cmd_snapshots.go
Comment (fd0) [inline cmd/restic/cmd_snapshots.go]: This is duplicate code, I would suggest extracting it into a function which takes a list of snapshots (`[]restic.Snapshot`, called a "slice" in Go) and returns a `map[string][]Snapshot` and an `error`. Let me know if you need help with this.

File: cmd/restic/cmd_snapshots.go
Comment (fd0) [inline cmd/restic/cmd_snapshots.go]: It's common in Go to return early, so the actual action in the function is not indented so far. How about this instead?
```suggestion
	if !GroupByTag && !GroupByHost && !GroupByPath {
		return nil
	}
```

File: cmd/restic/cmd_snapshots.go
Comment (fd0) [inline cmd/restic/cmd_snapshots.go]: Add so-called struct tags here to set the names in the JSON output
```suggestion
	GroupKey  groupKey `json:"group_key"`
```

File: cmd/restic/cmd_snapshots.go
Comment (fd0) [inline cmd/restic/cmd_snapshots.go]:
```suggestion
	Snapshots []Snapshot `json:"snapshots"`
```

File: cmd/restic/cmd_snapshots.go
Comment (fd0) [inline cmd/restic/cmd_snapshots.go]: Add so-called struct tags here to set the name for the JSON output (lowercase, snake_case)
```suggestion
	Hostname string `json:"hostname"`
```

**Review by fd0 — CHANGES_REQUESTED** [review]
Hey, I've had a look at the code, thanks for your contribution (especially as a first-time contributor new to Go)!

I've got a few things that you could improve in the code, but it looks great already.

 * The Tests on Travis fail because the file `internal/archiver/testing_test.go` is not formatted properly with `gofmt` (the Go formatting standard is an implementation, rather than a document), which has changed a bit recently in the Go 1.11 release. In order to pass the tests, you need to format the code with `gofmt` on Go 1.11.
 * There are modifications to files in `vendor`/, these should not be there. Did you maybe run `gofmt` on all files, including `vendor/`? Please remove the modifications from the PR.

For testing, we're using the builtin testing framework, you can easily run all tests with `go test ./...` in the repo, or chdir to some directory and just run `go test`.

Unfortunately, we don't have tests for user-interface interactions like this in place (at least not yet).

Let me know if you need any help!

File: cmd/restic/cmd_snapshots.go
Comment (ArcticXWolf) [reply cmd/restic/cmd_snapshots.go]: Okay, I extracted it into an own function. Is the style of the new function okay? And shall I remove the duplicate code (in the forget command) in this pull request or shall I open a new one after this got merged?

**Comment by ArcticXWolf** [comment]
Okay, this is nearly complete, I just have some last questions:

> For testing, we're using the builtin testing framework, you can easily run all tests with go test ./... in the repo, or chdir to some directory and just run go test.
> 
> Unfortunately, we don't have tests for user-interface interactions like this in place (at least not yet).

The tests succeed with my changes, but shall I add new ones? I have no idea how I should design such UI tests. Maybe just save the expected output into a fixture?

Furthermore I added a changelog file, whats missing now is the modification of the documentation. Where shall I mention it? Maybe at [the working with repos section](https://github.com/restic/restic/blame/master/doc/045_working_with_repos.rst#L56)?

File: cmd/restic/cmd_snapshots.go
Comment (ArcticXWolf) [reply cmd/restic/cmd_snapshots.go]: I put the shared code into the internal folder and made the forget command use it too.

**Comment by ArcticXWolf** [comment]
Okay, I refactored the forget command to use the shared code and added the documentation. Tests are green for me, just waiting for the CI to confirm this. As I do not know how to do the interface tests (and we did not have them before), I believe this is ready for review.

**Review by fd0 — APPROVED** [review]
LGTM, thanks! Sorry for the long delay before reviewing this!

**Comment by fd0** [comment]
I've rebased and fixed the conflict on `cmd_forget.go`, let's see what the tests say.

**Pull request merged**
